### PR TITLE
New scattering data format

### PR DESCRIPTION
**Pull request by simonpf**
This PR implements a new format for handling single scattering data.

The new format aims to address limitations of the current scattering-data handling:

- Allow for different incoming and scattering angle grids
- Allow representing scattering data using SHT coefficients
- Simplify handling of scattering data in ARTS

The PR is still a work in progress but I am opening it here to start a discussion on the integration into ARTS.

### CODE REVIEW

File: src/python_interface/py_scattering_new.cpp [hunk 1]
```diff
@@ -0,0 +1,244 @@
...
+
+  ////////////////////////////////////////////////////////////////////////////////
+  // Single scattering data
+  ////////////////////////////////////////////////////////////////////////////////
+
+  py::class_<scattering::SingleScatteringData>(m, "SingleScatteringData")
+      // Constructors
+      .def(py::init<>())
+      .def(py::init<const scattering::SingleScatteringData &>())
+      .def(py::init<scattering::math::Vector<double>,
```
[the remark this answers is not in the document]
Comment (simonpf) [reply src/python_interface/py_scattering_new.cpp:33]: The incorrect resolution of the Eigen types to numpy is due to a missing include so easy to fix.

I was not planning to expose SHT, so I will just remove the constructor.

File: src/scattering/array.h [hunk 2]
```diff
@@ -0,0 +1,84 @@
+#ifndef __ARTS_SCATTERING_ARRAY__
```
[the remark this answers is not in the document]
Comment (simonpf) [reply src/scattering/array.h:1]: Thanks for the info, will fix this.

File: src/scattering/particle.h [hunk 3]
```diff
@@ -0,0 +1,617 @@
...
+   *
+   * Linear interpolation of scattering data along frequency dimension.
+   *
+   * @param f_grid The frequency grid to interpolate the data to.
+   * @return New particle with the scattering data interpolated to the
+   * given frequencies.
+   */
+  Particle interpolate_frequency(
+      std::shared_ptr<math::Vector<double>> f_grid) const {
+    return Particle(properties_, data_.interpolate_frequency(f_grid));
```
[the remark this answers is not in the document]
Comment (simonpf) [reply src/scattering/particle.h:145]: If I understand the documentation of ``std::enable_shared_from_this`` correctly, I wouldn't help with the type signatures here. Moreover, isn't creating a shared pointer out of a reference just the misuse good example from the [docs](https://en.cppreference.com/w/cpp/memory/enable_shared_from_this)?

The motivation for using shared pointers here is actually two-fold:

1. This should save some memory since particles will finally be brought to the same grids, so it feels natural that particles on the same grid should share those vectors. I admit, however, that this is likely a case of premature optimization.
2. More importantly, using sharing the ownership makes it easier to discover if all data has been brought to the same grid.

[the remark this answers is not in the document]
Comment (riclarsson) [reply src/scattering/particle.h:145]: It is perhaps the misuse case.

I think I formulated myself poorly.  I understand that the `*this` class is never value constant, so the shared pointer may modify some of the underlying values.  Is it?

Because I feel like I want to interpolate to a grid or value, not to a pointer.  Sending in a shared pointer indicates I am not only interpolating but also doing something else.  So either the name of the function is wrong (perhaps it should be `interpolate_frequency_and_update_data`), or the function can take a constant reference to new frequency grid.

With the current name, and if you need to update the data but this will be done in such a way that the `Particle` that is returned is practically value-constant, using `enabled_shared_from_this` allows you to hide that detail while still keeping the name.  This was mostly to work around the name-understanding problem I had here

File: src/scattering/interpolation.h [hunk 4]
```diff
@@ -0,0 +1,791 @@
...
+
+namespace scattering {
+namespace detail {
+
+//
+// Interpolation helper.
+//
+
+template <typename Derived>
+struct is_vector {
```
[the remark this answers is not in the document]
Comment (simonpf) [reply src/scattering/interpolation.h:28]: I think that naming the struct ``is_compile_time_eigen_vector`` defeats the purpose of the ``detail`` namespace, so I will just leave that as it is.

File: src/scattering/particle.h (hunk 3 above)
[the remark this answers is not in the document]
Comment (riclarsson) [reply src/scattering/particle.h:145]: `enable_shared_from_this` would never be undefined behavior.  It always throws quite reliably.  Anyways, this is besides the point, it's fine you don't want to use it.

As I understand how this part of the code is intended to work without fault: you have some function that creates a bunch of `Particle`.  A single `Particle` needs to live on a frequency grid.  You might want many more `Particle` than frequency grids, so you let them share frequency grids.  This shared frequency grids are created as `const math::Vector<double>`, since you don't want it to change as you have already done the interpolation.  The original `const math::Vector<double>` pointers may then all live on after the function that created the `Particle`.

Now my understanding fails me: How can a shared pointer to a `math::Vector<double>` be guaranteed const valued?  If it is not const, then it seems you are at a risk of changing the grid values even though you've already interpolated the values.  This matters to me because there are two frequency grids in ARTS.  One sensor-frequency grid `f_grid` that users set, and then an internal Doppler shifted version that changes per atmospheric level due to wind.  It is the latter you likely want to use in these interpolations.

[the remark this answers is not in the document]
Comment (simonpf) [reply src/scattering/particle.h:145]: OK, so you just want me to the pointee vector ``const``?

File: src/scattering/arts_ssdb.cpp [hunk 5]
```diff
@@ -0,0 +1,659 @@
...
+namespace arts_ssdb {
+
+namespace detail {
+
+////////////////////////////////////////////////////////////////////////////////
+// Helper functions.
+////////////////////////////////////////////////////////////////////////////////
+
+std::pair<double, double> match_temp_and_freq(std::string group_name) {
+  std::regex group_regex(R"(Freq([0-9\.]*)GHz_T([0-9\.]*)K)");
```
[the remark this answers is not in the document]
Comment (simonpf) [reply src/scattering/arts_ssdb.cpp:22]: I don't think this is critical.

File: src/scattering/particle.h (hunk 3 above)
[the remark this answers is not in the document]
Comment (riclarsson) [reply src/scattering/particle.h:145]: If that is possible.  It allows all of my problems of understanding to go away :)

File: src/scattering/math.h [hunk 6]
```diff
@@ -0,0 +1,647 @@
...
+    return std::abs(a) < epsilon;
+}
+
+template <typename Scalar>
+Scalar save_acos(Scalar a, Scalar epsilon = 1e-6) {
+  if (equal(a, 1.0, epsilon)) {
+    return 0.0;
+  }
+  if (equal(a, -1.0, epsilon)) {
+    return M_PI;
```
[the remark this answers is not in the document]
Comment (simonpf) [reply src/scattering/math.h:644]: Wouldn't conversion operators defines in the ``scattering`` namespace make more sense? Then you avoid the coupling between the ``matpack*`` files and Eigen.

File: src/scattering/math.h [hunk 7]
```diff
@@ -0,0 +1,647 @@
+/** \file scattering/maths.h
```
[the remark this answers is not in the document]
Comment (simonpf) [reply src/scattering/math.h:1]: I always thought of the file as being named ``scattering/math.h``. But, yeah, I will rename it.

File: src/scattering/eigen_tensor.h [hunk 8]
```diff
@@ -0,0 +1,463 @@
...
+
+    Copyright (c) 2020 Simon Pfreundschuh <simon.pfreundschuh@chalmers.se>
+
+    All rights reserved. Use of this source code is governed by a
+    BSD-style license that can be found in the LICENSE file.
+*/
+#pragma once
+
+#include "pybind11/pybind11.h"
+#include "pybind11/numpy.h"
```
[the remark this answers is not in the document]
Comment (simonpf) [reply src/scattering/eigen_tensor.h:13]: Can you rephrase your concerns please.

File: src/scattering/eigen_tensor.h [hunk 9]
```diff
@@ -0,0 +1,463 @@
...
+//
+
+struct is_tensor_impl {
+    template <typename Scalar, int NumIndices, int Options, typename IndexType>
+        static std::true_type check(Eigen::Tensor<Scalar, NumIndices, Options, IndexType> *);
+    static std::false_type check(...);
+};
+
+template <typename T>
+using is_eigen_tensor = decltype(is_tensor_impl::check((intrinsic_t<T>*)nullptr));
```
[the remark this answers is not in the document]
Comment (simonpf) [reply src/scattering/eigen_tensor.h:62]: Since this is not a deal breaker, I will ignore this for now.

File: src/scattering/interpolation.h [hunk 10]
```diff
@@ -0,0 +1,791 @@
...
+ *
+ * Indirectly sorts elements in a vector and returns a vector of
+ * sorted indices.
+ *
+ * @param v: The vector to sort.
+ * @return A vector containing the indices that sort the given
+ *    vector into ascending order.
+ */
+template <typename Scalar>
+math::Vector<Eigen::Index> indirect_sort(const math::Vector<Scalar>& v) {
```
[the remark this answers is not in the document]
Comment (simonpf) [reply src/scattering/interpolation.h]: Moved it.

File: src/scattering/integration.h [hunk 11]
```diff
@@ -0,0 +1,584 @@
...
+   */
+  // pxx :: hide
+  void calculate_nodes_and_weights() {
+    const long int n = degree_;
+    const long int n_half_nodes = (n + 1) / 2;
+    const long int n_max_iter = 10;
+    Scalar x, x_old, p_l, p_l_1, p_l_2, dp_dx;
+
+    for (int i = 1; i <= n_half_nodes; ++i) {
+      p_l = M_PI;
```
[the remark this answers is not in the document]
Comment (simonpf) [reply src/scattering/integration.h:77]: I replaced ``M_PI`` with ``pi_v`` and added the ``std::floating_point`` concept to all ``Scalar`` template arguments.

File: src/scattering/eigen_tensor.h (hunk 8 above)
[the remark this answers is not in the document]
Comment (riclarsson) [reply src/scattering/eigen_tensor.h:13]: Yeah, a bit vague.

In short, I don't want `pybind11` in `artscore`.  These includes seem to define ways for you to interact with Eigen tensors using `pybind11` in C++, but it might be that the idea is that you just want these bindings in `pyarts`.  Are you planning to leak this part to `artscore` or have your C++ part depend on `pybind11`?  I would be against that leak.  If you do not plan to leak this to `artscore`, then I would like it if the file is renamed to `py_eigen_tensor.h`

[the remark this answers is not in the document]
Comment (simonpf) [reply src/scattering/eigen_tensor.h:13]: The header is only included by ``py_scattering_new.cpp``. So there's no leakage into ``artscore``. It's just an extension of the automatic interfacing capabilities of pybind11 provided by ``arts/scattering``.  Therefore, it makes sense to me to have the file located there, but if you have other preferences I am happy to move it.

[the remark this answers is not in the document]
Comment (riclarsson) [reply src/scattering/eigen_tensor.h:13]: I see.  Thanks, then I am happy with the file content.  I still want you to name it `py_eigen_tensor.h`, to be clearer with the intent.  It helps to keep the name clean, just as the `m_` names in `artscore` keeps the method files named cleanly.  @olemke , can I get a second opinion on naming policies here?

[the remark this answers is not in the document]
Comment (olemke) [reply src/scattering/eigen_tensor.h:13]: I agree. That's a good convention to follow.

File: src/scattering/math.h (hunk 6 above)
[the remark this answers is not in the document]
Comment (simonpf) [reply src/scattering/math.h]: Previously my approach was to define ``to_arts`` functions for the conversion. But your proposal will probably end up being cleaner. I just never liked the Tensor classes in ARTS and somehow hoped they'd die before I have to write proper conversion functions.

But what changes are you currently working on in #570? Should I wait for that?

[the remark this answers is not in the document]
Comment (riclarsson) [reply src/scattering/math.h]: I agree the Tensor classes are not nice but they are our core data communication unit.

You should wait.  I hope I fixed the lasts bits now so it is merged after lunch.

When this is merged, if you can make these Eigen tensors fit into the `tensorN_like` and `matrix_like` and `vector_like` concepts, the conceptual copy constructor should just work as it is.  You do this by making sure that you can call:

`library_size`, `vitrine_size`, `shelf_size`, `book_size`, `page_size`, `row_size`, and `column_size` with the Eigen tensor as input.

This in turn means you have to describe in these concepts how you reach that dimension via:
`library_keeper`, `vitrine_keeper`, `shelf_keeper`, `book_keeper`, `page_keeper`, `row_keeper`, and `column_keeper`

The advantage with concept is that things like `shelf_keeper<T>` evaluates to `true` or `false` in constexpr, so you can use that in the `*_size` functions to then call the appropriate member method.  I had a quick glance at the Eigen tensor class, and I think if you have some generic concept that goes something like this:

```cpp
template <typename T, std::size_t N>
concept eigen_tensor_of_minimum_size = T::rank >= N and requires(T a) {
  { a.dimension(N-1) } -> access_index;
};
```

in the `matpack::external_classes` namespace, then you should be able to just add `external_classes::eigen_tensor_of_minimum_size<T, 1>` to `column_keeper` and so on for the other `*_keeper`s.  Though I did not try this out myself.  I do not know, for instance, how you are supposed to access the `rank` actually.

**Comment by simonpf** [comment]
@olemke @riclarsson I would like to get your feedback on this, since you are more familiar with the ARTS build system.

I added the ``ScatteringSpecies`` class, which provides a generic scattering species for the use in the ARTS workspace. The generic functionality is defined by  the ``ScatteringSpeciesImpl`` abstract class, which which ``ScatteringSpecies`` interacts under the hood. The interface defined by ``ScatteringSpeciesImp`` defines methods that use the ``Workspace`` class. However,  I have to add the include of ``scattering_species.h`` to ``make_tokval_h.cc`` for  ``ScatteringSpecies`` to become an ARTS group. Ultimately, this creates a circular dependency because ``workspace_ng.h`` imports ``wsv_aux`` which in turn imports ``tokval.h``, which includes ``scattering_species.h`` which then tries to include ``workspace_ng.h``.

Is it the case that no ARTS group can make a reference to the Workspace class, or am I missing something here?

**Comment by riclarsson** [comment]
> Is it the case that no ARTS group can make a reference to the Workspace class, or am I missing something here?

This is the case.  We need to keep this separate as `TokVal` must be able to create any Arts Group, and `Agenda` being an Arts Group while also holding `TokVal` is not nice.  As far as I see for your code though, can't you just forward declare `Workspace`?  That is, just `class Workspace;` instead of `#include "workspace_ng.h"`.

**Comment by simonpf** [comment]
Thanks, @riclarsson. That works!

**Comment by simonpf** [comment]
@olemke @riclarsson  Can you maybe have  a look at the error occurring in the ubuntu-doc and ubuntu-gcc-12-debug builds?

Importing ``pyarts`` raises the following error:

````
ImportError: implicitly_convertible: Unable to find type ArrayOfScatteringSpecies
````

I have been dealing with this for some time but couldn't figure out the reason for this.

**Comment by riclarsson** [comment]
> ```
> ImportError: implicitly_convertible: Unable to find type ArrayOfScatteringSpecies
> ```

I think I found the error.  The culprit is line 1811 of `gen_auto_py.cpp`.  It demands that you have to be able to get any Arts group from a `WorkspaceVariable` type.  We need this or the entire `Workspace` integration fails (you cannot do ws.Print(ws.f_grid)` without it as `ws.Print` takes a `const Vector&`, not a `WorkspaceVariable`, which is what `ws.f_grid` returns).  You simply must have defined all Arts groups in python before you call into the automatic workspace integration.  That is, line 138 of `py_module.cpp`: `py_workspace(m, ws, wsv);`.

I cannot compile the code easily because it asks for an `fftw3.h` file.  I thought I installed this so I am not sure what the actual problem is; I even pass the CMake-tests.  Anyways, you can easily add the class by calling `PythonInterfaceWorkspaceArray(ScatteringSpecies)` at the end of `py_scattering_new.cpp` (if this is where `ScatteringSpecies` is defined).  This should solve the problem.  Note that you might want to ensure that the namespace is bare as a lot of documentation is auto-generated with that call.

I will have a look at the documentation.  If it is not stating that ALL Arts groups must be defined before the call to that line in `py_module.cpp`, then it must be fixed.

**Comment by olemke** [comment]
> I cannot compile the code easily because it asks for an `fftw3.h` file. I thought I installed this so I am not sure what the actual problem is; I even pass the CMake-tests.

True, cmake only checks if fftw is installed, but doesn't fail if it's not. @simonpf: A check needs to be added to `CMakelists.txt` to throw an error if fftw is not found. @riclarsson: You should be fine if you install fftw with `brew install fftw`.

**Comment by riclarsson** [comment]
> You should be fine if you install fftw with `brew install fftw`.

@olemke, I get both:
```bash
[cmake] -- Found FFTW: /opt/homebrew/lib/libfftw3.dylib  
```

and:
```bash
[ 46%] Building CXX object src/CMakeFiles/methods.dir/methods.cc.o
In file included from /Users/richard/Work/arts/src/scattering/scattering_data_field.h:29,
                 from /Users/richard/Work/arts/src/scattering/single_scattering_data_impl.h:16,
                 from /Users/richard/Work/arts/src/scattering/single_scattering_data.h:15,
                 from /Users/richard/Work/arts/src/scattering.h:44,
                 from /Users/richard/Work/arts/src/scattering_species.h:45,
                 from /Users/richard/Work/arts/build/src/tokval.h:27,
                 from /Users/richard/Work/arts/src/wsv_aux.h:31,
                 from /Users/richard/Work/arts/src/methods.cc:36:
/Users/richard/Work/arts/src/scattering/integration.h:9:10: fatal error: fftw3.h: No such file or directory
    9 | #include "fftw3.h"
      |          ^~~~~~~~~
compilation terminated.
make[2]: *** [src/CMakeFiles/methods.dir/methods.cc.o] Error 1
make[1]: *** [src/CMakeFiles/methods.dir/all] Error 2
make: *** [all] Error 2
```

Clearly I have it installed but not the headers in a path at a time they are required (methods.cc)

**Comment by riclarsson** [comment]
@simonpf It now imports.  You or @olemke must make netcdf a permanent ARTS dependency.  It took forever to recompile when `_nc_close` was not there :/

**Comment by simonpf** [comment]
Thanks! Yes, in retrospect that seems quite obvious. The user guide states

> Normally, there should be no problems adding new types and functions to the workspace
within the scope of the current implementations. The file gen_auto_py.cpp will gen-
erate the interface for you.

so I might have taken that a bit to literally.

**Comment by simonpf** [comment]
Currently using ``PythonInterfaceWorkspaceArray(ScatteringSpecies)`` won't work because ``ArrayOfScatteringSpecies`` is a subclass of ``Array<ScatteringSpecies>``.

How strongly do you object to that?

**Comment by riclarsson** [comment]
> Currently using `PythonInterfaceWorkspaceArray(ScatteringSpecies)` won't work because `ArrayOfScatteringSpecies` is a subclass of `Array<ScatteringSpecies>`.
> 
> How strongly do you object to that?

Not at all.  I am using the technique as well in `py_species.cpp` for `ArrayOfSpeciesTag`.  Take note when creating the stuff that you do the inheritance chaining.  That file was one of the first I added so it doesn't really do it justice.  But in short, `py::class_< ArrayOfScatteringSpecies, Array<ScatteringSpecies>>(...)`.  I think the `PythonInterfaceArrayDefault` can be implement only on the python class you define for `Array<ScatteringSpecies>` to give you the ability to convert to and from a python `list` among other things.

**Comment by simonpf** [comment]
@olemke @riclarsson Can you help me with the current build issues? It builds fine on my machine, so I am not sure what to do.

**Comment by olemke** [comment]
I've fixed the build issues with clang (at least with 15.0.6 locally). I hope they also work with the CI versions.
Feel free to yoink the commit [from my branch](https://github.com/olemke/arts/commit/24628a9502c9613b71e3fd0431432007fdd1df50).
Note that I also had to add the `FFTW_INCLUDE_DIR` to `src/CMakeLists.txt` since the `fftw3.h` is pulled in by `src/scattering.h`.
The GCC errors about `ENUMCLASS_EMPTY` are really strange. I cannot reproduce them either. But since `ENUMCLASS_EMPTY` was recently removed in master and `ENUMCLASS` handles both cases now, I suggest to first do another merge from master and see if that fixes the problem. Better to try that before spending more time on hunting down the problem in deprecated code.
